proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+  they do not need to know the details of how the model works or how to define tasks for it. This can
+  reduce the learning curve and make it easier for users to get started with MultiTaskComponent and
+  to leverage the power of LLMs in Haystack.
+
+
+- The extensible and modular approach of MultiTaskComponent allows users to easily add support for
+  additional tasks, even on-the-fly, which can enable them to extend the capabilities of the model and to use it for
+  a wider range of NLP tasks. This can provide users with more flexibility and control over the model,
+  and can enable them to tailor it to their specific needs and applications.
+
+  # Detailed design
+
+- The MultiTaskComponent class is the most important abstraction in Haystack's large language model (LLM) support.
+  In addition to MultiTaskComponent class, we'll also define a set of templates for the most popular NLP tasks.
+
+
+- NLP task templates will be represented by `TaskTemplate` dataclass class.
+
+  ``` python
+      @dataclass(frozen=True)
+      class TaskTemplate:
+
+          name: str
+          prompt_text: str
+          prompt_params: List[str]
+
+  ````
+
+  MultiTaskComponent defines 10-20 default NLP tasks with TaskTemplate instance list and allow users to register
+  additional tasks with MultiTaskComponent.
+
+
+- The default tasks (question-answering,question-generation, summarization etc.) could be examined by the user
+  using `supported_tasks_names` method. For example:
+
+  ``` python
+	  from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.supported_task_names()
+
+      ----------------------------------------------------------------------------
+      ['question-answering',
+       'question-generation',
+       'conditioned-question-generation',
+       'summarization',
+       'question-answering-check']
+  ```
+
+
+- MultiTaskComponent supports natural language prompting (using `prompt_for_task`) by specifying task in addition to
+  prompt. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+	  mtc.prompt_for_task("question-generation", documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- MultiTaskComponent supports selecting a particular current task (e.g. question-generation) and then subsequently
+  using the selected task until user changes the current task. For example:
+
+  ``` python
+	  qa = mtc.task("question-generation")
+      qa(documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- The addition of new tasks is supported by the `add_task` method. For example:
```

Comment:
In addition to this `add_task` method could we also consider allowing users to pass a `TaskTemplate` to the `mtc.prompt_for_task` method? So something like 
```python
task_template = TaskTemplate(name="sentiment-analysis",
                          prompt_text="Please give a sentiment for this context. Answer with positive, negative or neutral. Context: $documents; Answer:",
                          prompt_params=["documents"])
mtc.prompt_for_task(task_template, documents=["I really enjoyed the recent movie."])
```
This would allow users to not have to always register new tasks before being able to use them.

Reply:
+1

Reply:
Does this task then get registered as well? @sjrl

Reply:
Added to spec via https://github.com/deepset-ai/haystack/pull/3665/commits/6331ac2d2413eac956d7705dc9091a2f13eb314d Marking the comment resolved